src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -3762,25 +3782,42 @@ LogicalResult ONNXGreaterOrEqualOp::inferShapes(
       ONNXGreaterOrEqualOpAdaptor>(*this, b.getI1Type());
 }
 
+//===----------------------------------------------------------------------===//
+// ONNXHardmaxOp
+//===----------------------------------------------------------------------===//
+
 LogicalResult ONNXHardmaxOp::verify() {
-  ONNXHardmaxOpAdaptor hmOp = ONNXHardmaxOpAdaptor(*this);
-  auto input = hmOp.input();
-  int64_t axis = this->axis();
+  ONNXHardmaxOpAdaptor operandAdaptor(*this);
+  auto input = operandAdaptor.input();
+  if (!hasShapeAndRank(input))
+    // Won't be able to do any checking at this stage.
+    return success();
 
-  // Verify that axis must be in range [-r, r - 1], where r is the rank of
-  // input.
-  if (hasShapeAndRank(input)) {
-    int64_t rank = input.getType().cast<ShapedType>().getRank();
-    if (axis < -rank || axis > rank - 1)
-      return emitOpError("axis value is out of range");
-  }
+  // axis attribute must be in the range [-r,r-1], where r = rank(input).
+  int64_t axisValue = axis();
+  int64_t inputRank = input.getType().cast<ShapedType>().getRank();
+  if (axisValue < -inputRank || axisValue >= inputRank)
+    return onnx_mlir::Diagnostic::emitAttributeOutOfRangeError(
+        *this->getOperation(), "axis", axisValue,
+        onnx_mlir::Diagnostic::Range<int64_t>(-inputRank, inputRank - 1));
 
   return success();
 }
 
 LogicalResult ONNXHardmaxOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  getResult().setType(getOperand().getType());
+  auto inputType = input().getType().cast<ShapedType>();
```

Comment:
most inference code, including yours above start with this
```
  // Cannot infer the output shape if the input shape is not yet known.
   if (!hasShapeAndRank(input()))
     return success();
```
any reason this is missing here?

Reply:
I can add that, at some point the need to go through and the `inferShapes` member functions and make them all consistently use `hasShapeAndRank`.

Reply:
Please add this one here, as I think it is needed. And yes, one could go eventually through the code and update it, vouch for that too.

Reply:
done